sdk/search/Azure.Search.Documents/src/Batching/Publisher{T}.cs

```diff
@@ -0,0 +1,493 @@
+            {
+                await _channel.Writer.WriteAsync(message, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                // It's an unbounded channel so spinning should be fine
+                while (!_channel.Writer.TryWrite(message)) { /* spin */ }
+            }
+        }
+        #endregion
+
+        #region Message Consumer
+        /// <summary>
+        /// Listen for and process messages.
+        /// </summary>
+        /// <returns>
+        /// A Task that will run for the lifetime of the publisher.
+        /// </returns>
+        private async Task ProcessMessagesAsync()
+        {
+            CancellationToken cancellationToken = PublisherCancellationToken;
+            while (await _channel.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
+            {
+                Message message = await _channel.Reader.ReadAsync(cancellationToken).ConfigureAwait(false);
+                switch (message.Operation)
+                {
+                    case MessageOperation.Publish:
+                        await OnDocumentsAddedAsync(message.Documents, cancellationToken).ConfigureAwait(false);
+                        break;
+                    case MessageOperation.Flush:
+                        await OnFlushedAsync(cancellationToken).ConfigureAwait(false);
+                        break;
+                    default:
+                        throw new InvalidOperationException($"Unexpected value {message.Operation} of type {nameof(MessageOperation)}.");
+                }
+            }
+            StopTimer();
+        }
+
+        /// <summary>
+        /// Get a key used to identify a given document.
+        /// </summary>
+        /// <param name="document">The document.</param>
+        /// <returns>A key for that document.</returns>
+        protected abstract string GetDocumentKey(T document);
+
+        /// <summary>
+        /// Process any documents that have been added for publishing.
+        /// </summary>
+        /// <param name="documents">The documents to publish.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns>Task for processing the added documents.</returns>
+        protected virtual async Task OnDocumentsAddedAsync(IEnumerable<T> documents, CancellationToken cancellationToken)
+        {
+            // Add all of the documents to our queue
+            Debug.Assert(documents != null);
+            foreach (T document in documents)
+            {
+                PublisherAction<T> action = new PublisherAction<T>(document, GetDocumentKey(document));
+                _pending.Enqueue(action);
+            }
+
+            // Determine whether or not that crossed the threshold and we need
+            // to automatically submit the next batch
+            if (AutoFlush && HasFullBatch())
+            {
+                await PublishAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
It's because we're a single consumer processor - we don't want batches being submitted in parallel.  But thinking about this, I should only start the timer if we didn't AutoFlush.